### PR TITLE
[SDK] Refactoring swap-quote 

### PR DESCRIPTION
**Pull request by odcheung**
## Improvements
- More accurate quote values - ported the swap logic from contract & add comprehensive unit-test suite to confirm quote values are identical to contract results
- More info available in quote - endTickIndex, sqrtPrice, feeAmount etc...
- Additional swapQuoteFromOutputToken to allow users to ask for a quote (and swap input parameters) based on the desired receive amount
- Bug fix on TickArray calculation that led to the annoying 0x1787 (InvalidTickArraySequence) error
- Addresses all errors in the old-swap quote where it fails to traverse tick-arrays in many cases
- Bug fix on WhirlpoolClient.swap where it incorrectly uses the quote.estimated values rather than the user defined tokenAmount value

### Breaking API changes
- Parameter changes for the SwapQuote functions - `swapQuoteByInputToken`, `swapQuoteWithParams`
- Reorganized some swap specific methods in PoolUtils into a new SwapUtils class
  - Most notably, `PoolUtil.getTickArrayPublicKeysForSwap` is now `SwapUtils.getTickArrayPublicKeys`
  - To get the `TickArray` object that you can plug into the quote, use `SwapUtils.getTickArrays`

## Notes
- `anchor test` passes all unit-tests

### CODE REVIEW

**Review by philcchen — COMMENTED** [review]
Glossed a bit over the math/tests, assuming it's just a port; is there anything you'd want a deeper look at ?

File: sdk/src/quotes/public/swap-quote.ts [hunk 1]
```diff
@@ -1,484 +1,185 @@
...
-  const {
-    whirlpoolData: { tickSpacing },
-  } = baseInput;
-  return TickArrayUtil.getTickFromArray(tickArray, tickIndex, tickSpacing);
+/**
+ * Perform a sync swap quote based on the basic swap instruction parameters.
+ * @param params - SwapQuote parameters
+ * @returns a SwapQuote object with estimates on token amounts, fee & end whirlpool states.
+ */
+export function swapQuoteWithParams(params: SwapQuoteParam) {
```
[the remark this answers is not in the document]
Comment (odcheung) [reply sdk/src/quotes/public/swap-quote.ts:170]: We need to maintain a sync-version of the quote function in case power users already have their own flow in fetching data prior to quoting. (ex. Jupiter)

**Comment by odcheung** [comment]
> Glossed a bit over the math/tests, assuming it's just a port; is there anything you'd want a deeper look at ?

Glossing over the math sounds fine. We are already checking that the contract and quote are returning the same values in our test cases. I would say double-checking the tick-array-sequence (doing offset calculation rather than lazy-load traverse like our contract) and see if we need any more traverse test cases!

**Comment by rawfalafel** [comment]
Taking a look at the test that I believe you called out earlier: https://github.com/orca-so/whirlpools/blob/a4638f2bd254badeeef324ce65c603294522b98e/sdk/tests/sdk/whirlpools/swap/swap-traverse.test.ts#L171

My understanding is that if the current price is exactly on an initialized tick and the swap is moving left, the first tick array should be the tick array containing the current tick (in the test, array index 1): https://github.com/orca-so/whirlpools/blob/main/programs/whirlpool/src/state/tick.rs#L299-L307

However, in the SDK, the logic for fetching the tick array addresses ignores this rule and thinks that the first tick array account is the one below the tick array containing the current tick (in the test, array index 0): https://github.com/orca-so/whirlpools/blob/a4638f2bd254badeeef324ce65c603294522b98e/sdk/src/impl/whirlpool-impl.ts#L437-L440

This seems like an SDK bug and not a smart contract bug. Let me know if you think the analysis is correct. 

I also noticed a second test that's being ignored. Is this a potential bug too: https://github.com/orca-so/whirlpools/blob/a4638f2bd254badeeef324ce65c603294522b98e/sdk/tests/sdk/whirlpools/swap/swap-traverse.test.ts#L465

File: sdk/src/quotes/swap/tick-array-sequence.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+      }
+      return prev;
+    }, []);
+
+    // Edge case: nothing was ever touched.
+    if (result.length === 0) {
+      return [];
+    }
+
+    // If the result does not fit minArraySize, pad the rest with the last touched array
```
[the remark this answers is not in the document]
Comment (philcchen) [reply sdk/src/quotes/swap/tick-array-sequence.ts:61]: Seems a bit unideal, but 👍  Might be worth including this information in a comment here

File: sdk/src/types/public/constants.ts [hunk 3]
```diff
@@ -58,7 +59,19 @@ export const METADATA_PROGRAM_ADDRESS = new PublicKey(
...
  * @category Constants
  */
-export const MAX_TICK_ARRAY_CROSSINGS = 2;
+export const MAX_SWAP_TICK_ARRAYS = 3;
+
+/**
+ * The denominator which the protocol fee rate is divided on.
+ * @category Constants
+ */
+export const PROTOCOL_FEE_RATE_MUL_VALUE = new BN(10000);
```
Comment (philcchen) [inline sdk/src/types/public/constants.ts:71]: minor style: fwiw I find that when you have lots of 0, it can be nice to use the underscore syntax to make the numbers a bit more readable, i.e. `10_000` or `1_000_000`

**Review by philcchen — APPROVED** [review]
